### ISSUE TITLE
association scope (N:M) field always NULL

### ISSUE THREAD

**Issue by sharkovich** [sequelize/sequelize#5614]
Hello. I have set up N:M scoped association very similar to the example from [http://docs.sequelizejs.com/en/latest/docs/associations/#nm](documentation), and the scope field in my database is always empty. I am using Sequelize v3.19.3 with MySQL database (v5.6.28). Those are my  model definitions:

```
var Notification = sequelize.define("Notification", {
  ...
}, {
  classMethods: {
    associate: function (models) {
      Notification.belongsToMany(models.Post, {
        through: {
          model: models.NotificationTarget,
          unique: false,
        },
        foreignKey: {
          name: 'notificationToken',
        },
        constraints: false,
        as: 'TargetPost',
      });
      
      Notification.belongsToMany(models.Comment, {
        through: {
          model: models.NotificationTarget,
          unique: false,
        },
        foreignKey: {
          name: 'notificationToken',
        },
        constraints: false,
        as: 'TargetComment',
      });
      ...
    },
  },
});
```

```
var Post = sequelize.define('Post', {
  token: {
    type: DataTypes.STRING,
    primaryKey: true,
    unique: true,
  },
  ...
}, {
  classMethods: {
    associate: function (models) {
      ...
      Post.belongsToMany(models.Notification, {
        through: {
          model: models.NotificationTarget,
          unique: false,
          scope: {
            targetType: 'post',
          },
        },
        foreignKey: 'targetToken',
        constraints: false,
      });
    },
  },
});
```

```
var NotificationTarget = sequelize.define("NotificationTarget", {
  id: {
    type: DataTypes.INTEGER,
    autoIncrement: true,
    unique: true,
    primaryKey: true,
  },
  notificationToken: {
    type: DataTypes.STRING,
    unique: 'target_notification',
  },
  targetType: {
    type: DataTypes.STRING,
    unique: 'target_notification',
  },
  targetToken: {
    type: DataTypes.STRING,
    unique: 'target_notification',
    references: null,
  },
});
```

I create new notification instance and add a post to it:
```
db.Notification.create(data, {
  include: {
    all: true,
  },
}).then(function (notification) {
  if (post) {
    notification.addTargetPost(post);
  }
  notification.save();
});
```

I get no error whatsover, new records are created in both `notification` and `notification_target` tables, but `notification_target.targetType` field is always set to `NULL`. 

When i tried to get the notification from db:

```
db.Notification.findAll({
  where: {
    recipient: userToken,
  },
  include: [{
      model: db.Comment,
      as: 'TargetComment',
    }, {
      model: db.Post,
      as: 'TargetPost',
    }
  ]
});
```
for quite some time i was getting `Notification` object with associated Posts or Comments just as expected. Now on my local development setup botth `TargetPost` and `TargetComment` arrays are always empty. I haven't made any changes, and in test enviorment i still get associated records just fine.

Insert log:
```
Executing (default): INSERT INTO `tb_notifications` (`token`,`isRead`,`type`,`sender`,`recipient`,`createdAt`,`updatedAt`) VALUES ('MMpB39MBoRE5',false,'has-liked','8BxmIea1WIOJ','2DegpO','2016-03-21 12:02:17','2016-03-21 12:02:17');
Executing (default): INSERT INTO `tb_notifications_target` (`id`,`notificationToken`,`targetToken`,`createdAt`,`updatedAt`) VALUES (NULL,'MMpB39MBoRE5','vretxC','2016-03-21 12:02:17','2016-03-21 12:02:17');
```

Select log:
```
Executing (default): SELECT `Notification`.`token`, `Notification`.`isRead`, `Notification`.`type`, `Notification`.`sender`, `Notification`.`recipient`, `Notification`.`createdAt`, `Notification`.`updatedAt`, `Sender`.`username` AS `Sender.username`, `Sender`.`name` AS `Sender.name`, `Sender`.`avatar` AS `Sender.avatar`, `Sender`.`email` AS `Sender.email`, `Sender`.`token` AS `Sender.token`, `TargetComment`.`token` AS `TargetComment.token`, `TargetComment`.`content` AS `TargetComment.content`, `TargetComment`.`deleted` AS `TargetComment.deleted`, `TargetComment`.`createdAt` AS `TargetComment.createdAt`, `TargetComment`.`updatedAt` AS `TargetComment.updatedAt`, `TargetComment`.`postToken` AS `TargetComment.postToken`, `TargetComment`.`userToken` AS `TargetComment.userToken`, `TargetComment`.`parentToken` AS `TargetComment.parentToken`, `TargetComment.NotificationTarget`.`id` AS `TargetComment.NotificationTarget.id`, `TargetComment.NotificationTarget`.`notificationToken` AS `TargetComment.NotificationTarget.notificationToken`, `TargetComment.NotificationTarget`.`targetType` AS `TargetComment.NotificationTarget.targetType`, `TargetComment.NotificationTarget`.`targetToken` AS `TargetComment.NotificationTarget.targetToken`, `TargetComment.NotificationTarget`.`createdAt` AS `TargetComment.NotificationTarget.createdAt`, `TargetComment.NotificationTarget`.`updatedAt` AS `TargetComment.NotificationTarget.updatedAt`, `TargetPost`.`id` AS `TargetPost.id`, `TargetPost`.`token` AS `TargetPost.token`, `TargetPost`.`published` AS `TargetPost.published`, `TargetPost`.`title` AS `TargetPost.title`, `TargetPost`.`canonicalTitle` AS `TargetPost.canonicalTitle`, `TargetPost`.`excerpt` AS `TargetPost.excerpt`, `TargetPost`.`createdAt` AS `TargetPost.createdAt`, `TargetPost`.`updatedAt` AS `TargetPost.updatedAt`, `TargetPost`.`deletedAt` AS `TargetPost.deletedAt`, `TargetPost`.`language` AS `TargetPost.language`, `TargetPost`.`status` AS `TargetPost.status`, `TargetPost`.`userToken` AS `TargetPost.userToken`, `TargetPost`.`coverPhoto` AS `TargetPost.coverPhoto`, `TargetPost`.`iconPhoto` AS `TargetPost.iconPhoto`, `TargetPost.NotificationTarget`.`id` AS `TargetPost.NotificationTarget.id`, `TargetPost.NotificationTarget`.`notificationToken` AS `TargetPost.NotificationTarget.notificationToken`, `TargetPost.NotificationTarget`.`targetType` AS `TargetPost.NotificationTarget.targetType`, `TargetPost.NotificationTarget`.`targetToken` AS `TargetPost.NotificationTarget.targetToken`, `TargetPost.NotificationTarget`.`createdAt` AS `TargetPost.NotificationTarget.createdAt`, `TargetPost.NotificationTarget`.`updatedAt` AS `TargetPost.NotificationTarget.updatedAt` FROM `tb_notifications` AS `Notification` LEFT OUTER JOIN `tb_users` AS `Sender` ON `Notification`.`sender` = `Sender`.`token` AND `Sender`.`deletedAt` IS NULL LEFT OUTER JOIN (`tb_notifications_target` AS `TargetComment.NotificationTarget` INNER JOIN `tb_comments` AS `TargetComment` ON `TargetComment`.`token` = `TargetComment.NotificationTarget`.`targetToken`) ON `Notification`.`token` = `TargetComment.NotificationTarget`.`notificationToken` LEFT OUTER JOIN (`tb_notifications_target` AS `TargetPost.NotificationTarget` INNER JOIN `tb_posts` AS `TargetPost` ON `TargetPost`.`token` = `TargetPost.NotificationTarget`.`targetToken`) ON `Notification`.`token` = `TargetPost.NotificationTarget`.`notificationToken` AND `TargetPost`.`deletedAt` IS NULL WHERE `Notification`.`recipient` = '2DegpO' AND `Notification`.`token` = 'MMpB39MBoRE5';
```

Any help and/or corrections would be very much appreciated.

**Comment by mickhansen** [comment]
The scope only works when calling from the side that has the scope configured, post in this case.
Try using `post.addNotification(notification)`

**Comment by sharkovich** [comment]
@mickhansen thank you, it is was not really obvious for me from reading documentation.

But still, after making a change, even that `notification_target.targetType` field was set correctly, still `TargetPost` array was empty even if correct records were present in database. After a little bit of tinkering and analyzing queries sent by Sequelize, I have found out that includes need `required` option set to `true`.

**TL;DR:** 
Following code works flawlessly
```
db.Notification.create(data, {
  include: {
    all: true,
  },
}).then(function (notification) {
  if (post) {
    post.addNotification(notification);
  }
  notification.save();
});
```
```
db.Notification.findAll({
  where: {
    recipient: userToken,
  },
  include: [{
      model: db.Comment,
      as: 'TargetComment',
      required: true,
    }, {
      model: db.Post,
      as: 'TargetPost',
      required: true,
  }]
});
```

**Comment by mickhansen** [comment]
Correct, if the main model needs to be filtered by the include you need `required: true`
